src/client/Microsoft.Identity.Client/Instance/AuthorityEndpoints.cs

```diff
@@ -40,13 +42,26 @@ private async static Task UpdateAuthorityWithPreferredNetworkHostAsync(Authentic
         {
             InstanceDiscoveryMetadataEntry metadata = await
                 requestParameters.RequestContext.ServiceBundle.InstanceDiscoveryManager.GetMetadataEntryAsync(
-                    requestParameters.AuthorityInfo.CanonicalAuthority,
-                    requestParameters.RequestContext)
+                    requestParameters.WithAzureRegion ? 
+                        await buildAuthorityWithRegionAsync(requestParameters.AuthorityInfo).ConfigureAwait(false) : 
+                        requestParameters.AuthorityInfo.CanonicalAuthority,
+                    requestParameters.RequestContext,
+                    requestParameters.WithAzureRegion)
                 .ConfigureAwait(false);
 
             requestParameters.Authority = Authority.CreateAuthorityWithEnvironment(
                     requestParameters.AuthorityInfo,
                     metadata.PreferredNetwork);
         }
+
+        private async static Task<string> buildAuthorityWithRegionAsync(AuthorityInfo authorityInfo)
```

Comment:
Did you explore the possibility of putting this logic in the DIscoveryManager itself? It could be a DiscoveryProvider on its own.

Reply:
Done.